check-dispatcher refactor and es6 conversion

GHaberis opened the pull request:
## What does this change?

- Converts check-dispatcher to es6.

- Moves 'dependent check' logic out of check-mediator-checks and into check-dispatcher. Moving this logic into single place (check-dispatcher) makes reasoning about what checks are doing easier, rather than splitting logic across multiple files.

- Updates outbrain.js to run 3 checks isUserInContributionsAbTest, isUserInEmailAbTestAndEmailCanRun and isStoryQuestionsOnPage individually rather than running single test isOutbrainNonCompliant. Running these checks individually means we can track the actual cause of a non-compliant outbrain widget more granularly.

- Update outbrain-load.js to track state of outbrain component.

 What is the value of this and can you measure success?

- More es6
- Consolidates check-mediator/dispatcher logic into single location
- Simplifies check-mediator as no longer needs to manage dependent checks
- Allows us to track outbrain state more granularly

## Does this affect other platforms - Amp, Apps, etc?

No

## Screenshots

N/A

## Tested in CODE?

No

GHaberis commented:
@guardian/commercial-dev I've made some updates to the check-mediator/dispatcher that determine the state of outbrain in this PR. The decision making logic remains the same, where it sits has moved around a little, it's hopefully easier to follow now. The changes made here will allow us to track what condition/check resulted in a non-compliant outbrain widget at a more granular level.

SiAdcock commented on static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain-tracking.js:
```diff
@@ -2,10 +2,10 @@
-const tracking = function(trackingObj: {
-    widgetId?: string,
+const tracking = (trackingObj: {
+    widgetId?: ?string,
```
Since this field is already optional, I don't think it makes sense for it to be nullable as well. Instead of explicitly setting this to `null`, why don't we simply omit the property altogether? Does tracker-js (or Ophan itself) behave differently depending on whether this property is `null` or `undefined`?

GHaberis replied:
@SiAdcock yes Ophan itself behaves differently if widgetId isn't present, it ignores requests without widgetId right now. 

In order to track the info we need I'm going to need to raise a PR against ophan/ophan-data-lake to handle widgetId as null but yes it's probably cleaner to make it optional. 

I'll get the PR in order for this repos before merging this.

SiAdcock commented on static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.spec.js:
```diff
@@ -83,115 +80,121 @@ describe('Outbrain', () => {
+            resolveCheck('isStoryQuestionsOnPage', false);
+
+            init().then(() => {
+                expect(load).toHaveBeenCalled();
```
Seems redundant, given the next line 😄

GHaberis replied:
Can't really argue with this!

SiAdcock commented on static/src/javascripts/projects/common/modules/check-dispatcher.js:
```diff
@@ -0,0 +1,166 @@
+import { checks } from './check-mediator-checks';
+import { resolveCheck, waitForCheck } from './check-mediator';
+
+const someCheckPassed = results => results.includes(true);
```
Are we going to add Flow type annotations as part of this conversion?

GHaberis replied:
I wasn't sure what the suggested practise was regarding types, I followed the approach that they're mandatory on public functions then optional on private functions? I'm happy to add type annotations if you'll feel they'll provide useful guidance for developers editing/reading this file!

SiAdcock replied:
As we discussed IRL, I don't think there's any distinction in type annotation requirements between public and private functions. It's a case of balancing time you want to spend versus the [piece of mind that Flow gives us](https://github.com/guardian/frontend/wiki/So-you-want-to-ES6%3F#specific-way) 😅

gustavpursche commented on static/src/javascripts/projects/common/modules/check-dispatcher.js:
```diff
@@ -0,0 +1,166 @@
+
+        return Promise.all(dependentChecks).then(results =>
+            everyCheckPassed(results)
+        );
```
`return Promise.all(dependentChecks).then(everyCheckPassed);`. For the following occurences too.

regiskuckaertz commented:
@SiAdcock @gustavpursche Are we allowed to use `async`/`await`? Part of the reason why this module is hard to read is the promise chain, and I wonder if it would make our lives easier if it was written in imperative style. Not sure I got this right, but my attempt [here](https://gist.github.com/regiskuckaertz/3c34632d36874614e485efc8a74c5e6a).

SiAdcock commented:
@regiskuckaertz 

From what I understand, in order to transpile async functions, we would need to use Babel's [`transform-async-to-generator`](https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-async-to-generator) plugin. In turn, this requires us to serve a [generator runtime](https://github.com/facebook/regenerator/blob/master/packages/regenerator-runtime/runtime.js) (or [Babel's polyfill library](https://babeljs.io/docs/usage/polyfill/)), which adds a lot of weight to the application.

We could potentially spike out how much extra weight this would add. Or @sndrs may be able to shed further light on this subject.

PRBuilds commented:
PRbuilds results:


**Screenshots**
[mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16864/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png) • [wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16864/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16864/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16864/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png)



**Exceptions** (0)
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16864/thrown-exceptions.js)





**A11y validation** (0)
[a11y-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16864/a11y-report.txt)


--automated message

GHaberis commented:
@SiAdcock I've removed the references to 'state' in the tracking calls as the Ophan work to support this new property needs to be done before this data can be recorded. 

Rather than blocking this PR I'll raise a new PR to add 'state' to the tracking calls once Ophan is ready to accept it.

SiAdcock commented on static/src/javascripts/projects/commercial/modules/third-party-tags/outbrain.js:
```diff
@@ -1,84 +1,95 @@
 */
-const canLoadInstantly = function() {
-    return detect.adblockInUse.then(
+const canLoadInstantly = () =>
```
How do you feel about adding return / param types to these functions?

GHaberis replied:
Yep i'm happy to do this 👍

SiAdcock commented on static/src/javascripts/projects/common/modules/check-dispatcher.js:
```diff
@@ -0,0 +1,170 @@
+        return config.switches.emailInArticleOutbrain;
+    },
+
+    hasHighPriorityAdLoaded(): any {
```
Just noticed the signature of this and `hasLowPriorityAdLoaded` are inconsistent with other asynchronous checks in here. It might be dangerous to sometimes return a promise and sometimes not, as the calling function can't rely on the return value being `thennable`. Is it possible to make these function always return a `Promise<boolean>`, for safety?

SiAdcock replied:
To go further, perhaps all checks should always return `Promise<boolean>` even if they are not async, for consistency?

GHaberis replied:
The return values of these checks don't have to be `thennable`, their return value returned is passed as an argument to the `resolve` function for the promises defined in check-mediator. These promises can be resolved with a boolean or another promise.

SiAdcock commented on static/src/javascripts/projects/common/modules/check-mediator.js:
```diff
@@ -90,17 +51,16 @@ const waitForCheck = (id: string): Promise<any> => {
 
+const initCheckMediator = (): void => {
+    checks.forEach(registerCheck);
+};
```
If this function initially set `registeredChecks = {}`, we wouldn't need to expose the `testClean` function 🎉

gustavpursche replied:
Nice idea!

GHaberis commented:
@SiAdcock I've addressed these outstanding issues now.

- added types to check-dispatcher
- return type of all checks in dispatcher is now a Promise
- remove _ test function in check-mediator.
